Zend/zend_observer.c

```diff
@@ -38,12 +38,19 @@ typedef struct _zend_observer_fcall_data {
 	zend_observer_fcall_handlers handlers[1];
 } zend_observer_fcall_data;
 
+typedef struct _zend_observer_frame {
+	zend_execute_data *execute_data;
+	struct _zend_observer_frame *prev;
+} zend_observer_frame;
```

Comment:
I don't understand the purpose of this. Why can't we walk the current_execute_data at the time of bailout backwards?

Reply:
If we walked the stack right before the [current_execute_data gets set to NULL](https://heap.space/xref/php-src/Zend/zend.c?r=da0663a3#1155), then all the end handlers will fire prematurely if the zend_bailout is "caught".

Instead of tracking each observed frame, we could just keep `current_observed_frame` as a pointer to the last observed execute_data, and walk the stack backward from that point during request shutdown. WDYT?

Reply:
> Instead of tracking each observed frame, we could just keep current_observed_frame as a pointer to the last observed execute_data, and walk the stack backward from that point during request shutdown. WDYT?

That sounds reasonable to me.

Reply:
@nikic I [gave this approach a shot](https://github.com/php/php-src/commit/a00463ad58a6b857a13c1ee901a26a3f2898af31); when a function call/include is finished, the end handler sets the `current_observed_frame` to the `prev_execute_data`. This works fine for all cases except for file includes.

Once the `current_observed_frame` is set to a file include on the last observed end handler (during a non-zend_bailout "happy path"), the op_array for that include is destroyed before `php_request_shutdown` is called. Then `zend_observer_fcall_end_all` tries to obtain the map pointer from the `current_observed_frame`'s op_array that has already been freed.

Can you think of a different approach to address the issue of freed op_arrays, or should we go ahead with the existing solution?

Reply:
Does this affect only the top-level file or intermediate includes?

Reply:
It appears to be just the top-level file actually. I tried digging into it more deeply to figure out exactly how intermediate includes don't seem to have this issue. I didn't figure it out exactly, but it looks like perhaps the runtime cache thats pointed to from the execute_data is preserved even after the op_array is freed?

Reply:
I finally figured it out. We just needed to keep track of the first observed frame. Fixed in 8877bef. Let me know if you think this looks good for merge now. :)